File: backend/src/main/java/com/telehealth/pro/service/CloudinaryService.java
package com.telehealth.pro.service;

import com.cloudinary.Cloudinary;
import com.cloudinary.utils.ObjectUtils;
import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;
import java.util.UUID;

@Service
public class CloudinaryService {

    @Value("${cloudinary.cloud-name}")
    private String cloudName;

    @Value("${cloudinary.api-key}")
    private String apiKey;

    @Value("${cloudinary.api-secret}")
    private String apiSecret;

    @Value("${server.port:5000}")
    private String port;

    private Cloudinary cloudinary;
    private boolean useCloudinary;

    @PostConstruct
    public void init() {
        useCloudinary = cloudName != null && !cloudName.isEmpty() && !cloudName.equals("mock_cloudinary_cloud_name");
        if (useCloudinary) {
            cloudinary = new Cloudinary(ObjectUtils.asMap(
                    "cloud_name", cloudName,
                    "api_key", apiKey,
                    "api_secret", apiSecret
            ));
            System.out.println("[CLOUDINARY] Configured with real credentials.");
        } else {
            System.out.println("Cloudinary credentials missing or default. Local fallback will be used for file uploads.");
        }
    }

    public UploadResult uploadFile(MultipartFile file, String folder) throws IOException {
        if (useCloudinary) {
            try {
                Map<?, ?> uploadResult = cloudinary.uploader().upload(file.getBytes(), ObjectUtils.asMap(
                        "folder", folder,
                        "resource_type", "auto"
                ));
                return UploadResult.builder()
                        .url((String) uploadResult.get("secure_url"))
                        .publicId((String) uploadResult.get("public_id"))
                        .build();
            } catch (Exception e) {
                System.err.println("[CLOUDINARY ERROR] Upload failure: " + e.getMessage());
                throw new IOException("Cloudinary upload failed", e);
            }
        } else {
            try {
                String uploadDirPath = "public/uploads";
                File uploadDir = new File(uploadDirPath);
                if (!uploadDir.exists()) {
                    uploadDir.mkdirs();
                }

                String originalFilename = file.getOriginalFilename();
                String ext = "";
                if (originalFilename != null && originalFilename.contains(".")) {
                    ext = originalFilename.substring(originalFilename.lastIndexOf("."));
                }
                String newFilename = UUID.randomUUID().toString() + ext;
                Path destination = Paths.get(uploadDirPath, newFilename);

                Files.write(destination, file.getBytes());

                String fileUrl = "http://localhost:" + port + "/uploads/" + newFilename;
                return UploadResult.builder()
                        .url(fileUrl)
                        .publicId(newFilename)
                        .build();
            } catch (Exception e) {
                System.err.println("[LOCAL UPLOAD ERROR] Fallback upload failure: " + e.getMessage());
                throw new IOException("Local upload fallback failed", e);
            }
        }
    }

    public void deleteFile(String publicId) {
        if (useCloudinary) {
            try {
                cloudinary.uploader().destroy(publicId, ObjectUtils.emptyMap());
                System.out.println("[CLOUDINARY] File deleted: " + publicId);
            } catch (Exception e) {
                System.err.println("[CLOUDINARY ERROR] Delete failure: " + e.getMessage());
            }
        } else {
            try {
                Path filePath = Paths.get("public/uploads", publicId);
                Files.deleteIfExists(filePath);
                System.out.println("[LOCAL FILE] Static upload deleted: " + publicId);
            } catch (Exception e) {
                System.err.println("[LOCAL DELETE ERROR] Delete failure: " + e.getMessage());
            }
        }
    }
}

File: backend/bin/src/services/cloudinary.service.js
const cloudinary = require('cloudinary').v2;
const fs = require('fs');
const path = require('path');

const useCloudinary = process.env.CLOUDINARY_CLOUD_NAME && 
                      process.env.CLOUDINARY_CLOUD_NAME !== 'mock_cloudinary_cloud_name';

if (useCloudinary) {
  cloudinary.config({
    cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
    api_key: process.env.CLOUDINARY_API_KEY,
    api_secret: process.env.CLOUDINARY_API_SECRET
  });
} else {
  console.log('Cloudinary credentials missing or default. Local fallback will be used for file uploads.');
}

const uploadFile = async (localFilePath, folder = 'telehealth') => {
  if (useCloudinary) {
    try {
      const result = await cloudinary.uploader.upload(localFilePath, {
        folder: folder,
        resource_type: 'auto'
      });
      // Delete temporary local file
      if (fs.existsSync(localFilePath)) {
        fs.unlinkSync(localFilePath);
      }
      return {
        url: result.secure_url,
        publicId: result.public_id
      };
    } catch (error) {
      console.error('Cloudinary upload failure:', error);
      throw error;
    }
  } else {
    // Local fallback: move file to static upload folder
    try {
      const uploadDir = path.join(__dirname, '..', '..', 'public', 'uploads');
      if (!fs.existsSync(uploadDir)) {
        fs.mkdirSync(uploadDir, { recursive: true });
      }

      const fileName = path.basename(localFilePath);
      const destination = path.join(uploadDir, fileName);

      fs.copyFileSync(localFilePath, destination);
      if (fs.existsSync(localFilePath)) {
        fs.unlinkSync(localFilePath);
      }

      // Return a relative path URL or a local server URL
      const port = process.env.PORT || 5000;
      return {
        url: `http://localhost:${port}/uploads/${fileName}`,
        publicId: fileName
      };
    } catch (error) {
      console.error('Local fallback upload failure:', error);
      throw error;
    }
  }
};

const deleteFile = async (publicId) => {
  if (useCloudinary) {
    try {
      await cloudinary.uploader.destroy(publicId);
      return { success: true };
    } catch (error) {
      console.error('Cloudinary delete failure:', error);
      throw error;
    }
  } else {
    try {
      const uploadDir = path.join(__dirname, '..', '..', 'public', 'uploads');
      const filePath = path.join(uploadDir, publicId);
      if (fs.existsSync(filePath)) {
        fs.unlinkSync(filePath);
      }
      return { success: true };
    } catch (error) {
      console.error('Local file delete failure:', error);
      return { success: false, error: error.message };
    }
  }
};

module.exports = {
  uploadFile,
  deleteFile
};
